### DISCUSSION TITLE
Question: What limits lwip usage within a FreeRTOS environment to setup/loop and not tasks?

### DISCUSSION THREAD

**Discussion by spiderkeys** [earlephilhower/arduino-pico#2152] [General]
I was wondering if I could get a bit of clarification/context on the following lines in the Pico core's documentation:

WiFi docs:
> Multicore is supported, but only core 0 may run WiFi related code.
> FreeRTOS is supported only on core 0 and from within setup and [``](https://arduino-pico.readthedocs.io/en/latest/wifi.html#id1)loop`, not tasks, due to the requirement for a very different LWIP implementation. PRs always appreciated!

Ethernet Docs:
> The same restrictions for WiFi apply to these Ethernet classes, namely:
> Only core 0 may run any networking related code.
> In FreeRTOS, only the setup and loop task can call networking libraries, not any tasks.

I can understand why networking code might need to be restricted to a single core, but I am not clear on why networking code can not be called within a FreeRTOS task other than the one that calls setup/loop.

I've been looking at how the [zenoh-pico](https://github.com/eclipse-zenoh/zenoh-pico) middleware library works with the ESP32 Arduino core, and they seem to be able to make calls to lwip functions from within additional FreeRTOS tasks (pinned to core 0). I was wondering if I could get zenoh-pico working on the rp2040, but the limitation as written suggests that zenoh's background network tasks ([read and lease tasks](https://github.com/eclipse-zenoh/zenoh-pico/blob/main/examples/arduino/z_sub.ino#L85), which you normally create in `setup()` and call lwip's socket functions) would be in violation of this constraint.

Any background or clarifications are appreciated - I might be reading this entirely wrong!

**Comment by earlephilhower** [comment]
We run in LWIP's NO_SYS=1 mode which lets us run without an underlying OS.   LWIP and the WiFi libraries are non-reentrant, so having more than 1 task accessing them will make your app very unhappy.

But, more fundamentally, by running LWIP in NO_SYS mode there are no `socket`s at all.  So, the app you're talking about would need to be rewritten from scratch to use raw LWIP calls and not the POSIX emulation layer.

**Reply by spiderkeys** [reply]
Thanks for the context - it makes sense now, considering LWIP is set to NO_SYS=1.
